### ISSUE TITLE
ef core 3.X regression - left join + include exclude some results

### ISSUE THREAD

**Issue by Primith** [dotnet/efcore#21527]
When using in the same query DefaultIfEmpty and Include, some results are missing.

I reproduce this example in LinqPad with the Northwind Database : 

```C#
var joinQuery = from cust in Customers
join order in Orders.Include(f => f.OrderDetails)
  on cust.CustomerID equals order.CustomerID into group1
from order in group1.DefaultIfEmpty()
where order.ShipCountry == "France"
select new { cust, order };
                           
joinQuery.ToList().Dump();
joinQuery.Count().Dump();
```

With 2.2.6, I have 77 items in my list et 77 as result of the count query.
With 3.1.3 (or all other versions >= 3.0), I have 10 results in my list et 77 as result of the count query.

If I remove the left join clause, I have 77 results : 
```C#
var joinQuery = from cust in Customers
 join order in Orders.Include(f => f.OrderDetails)
       on cust.CustomerID equals order.CustomerID //into group1
//from order in group1.DefaultIfEmpty()
where order.ShipCountry == "France"
select new { cust, order };
                           
joinQuery.ToList().Dump();
joinQuery.Count().Dump();
```

If I remove the include, I have 77 results :
```C#
var joinQuery = from cust in Customers
join order in Orders//.Include(f => f.OrderDetails)
     on cust.CustomerID equals order.CustomerID into group1
from order in group1.DefaultIfEmpty()
where order.ShipCountry == "France"
select new { cust, order };
                           
joinQuery.ToList().Dump();
joinQuery.Count().Dump();
```

But if I use include + left join clause, some results are excluded.

### Further technical details

EF Core version: 3.X vs 2.2.6
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: ef core 3.1.5
Operating system: Windows
IDE: LinqPad, VS code

**Comment by Primith** [comment]
Hello,

I'm little worried about this issue. My project has started a few month ago and acceptance tests are failing due to this issue. From my point of view, it's only a simple SELECT request.

Are my requests so strange that I'm the only one to have this behaviour or do I badly use ef  ?

For information, for my real request, I have only 31 000 results whereas I expect 55 000.

Have I to give up entity framework and find a bypass with an other ORM or use stored procedure ? I have to know in order to decide if I continue my project with Entity framework or make a changement from now.

Thanks for your reply.

**Comment by smitpatel** [comment]
This issue existed in 3.1.x release. It has been fixed in 5.0-preview6 release.

**Comment by Primith** [comment]
So, if I understand correctly, you confirm that this issue is a regression from 3.1.x releases and the only response is to use a preview release for 5.0 which is not a LTS release whereas the 3.1 is.

According to https://docs.microsoft.com/fr-fr/ef/core/what-is-new/release-planning :
_We are more likely to patch an issue if:_

- _It is impacting multiple customers_

- _It is a regression from a previous release_
- _The failure causes data corruption_

I think I'm not the only customer with this issue.
For the regression part, the debate is closed.
The failure causes data corruption when reading, the results being totally wrong.

So, can you please study to report fix on 3.1.X next release please ?

Thanks in advance.

**Comment by smitpatel** [comment]
> The fix presents a high risk of breaking something else

We are less likely to fix this but I will reopen to discuss with the team.

**Comment by ajcvickers** [comment]
We discussed this as a team, but unfortunately the fix is too risky to port back to 3.1.
